ecosystem/typescript/sdk/src/utils/memoize-decorator.ts

```diff
@@ -0,0 +1,144 @@
+/**
+ * Credits to https://github.com/darrylhodgins/typescript-memoize
+ */
+
+/* eslint-disable no-param-reassign */
+/* eslint-disable no-restricted-syntax */
+
+interface MemoizeArgs {
+  expiring?: number;
+  hashFunction?: boolean | ((...args: any[]) => any);
+  tags?: string[];
+}
+
+export function Memoize(args?: MemoizeArgs | MemoizeArgs["hashFunction"]) {
+  let hashFunction: MemoizeArgs["hashFunction"];
+  let duration: MemoizeArgs["expiring"];
+  let tags: MemoizeArgs["tags"];
+
+  if (typeof args === "object") {
+    hashFunction = args.hashFunction;
+    duration = args.expiring;
+    tags = args.tags;
+  } else {
+    hashFunction = args;
+  }
+
+  return (target: Object, propertyKey: string, descriptor: TypedPropertyDescriptor<any>) => {
+    if (descriptor.value != null) {
+      descriptor.value = getNewFunction(descriptor.value, hashFunction, duration, tags);
+    } else if (descriptor.get != null) {
+      descriptor.get = getNewFunction(descriptor.get, hashFunction, duration, tags);
+    } else {
+      throw new Error("Only put a Memoize() decorator on a method or get accessor.");
+    }
+  };
+}
+
+export function MemoizeExpiring(expiring: number, hashFunction?: MemoizeArgs["hashFunction"]) {
```

Comment:
This name is okay though, I just thing the argument `expiring` is very vague vs `durationSecs` or `expirationUnixtime` or something.

Reply:
expiring is renamed